src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1090,6 +1090,14 @@ bool PhaseIdealLoop::create_loop_nest(IdealLoopTree* loop, Node_List &old_new) {
     if (UseProfiledLoopPredicate) {
       add_parse_predicate(Deoptimization::Reason_profile_predicate, inner_head, outer_ilt, cloned_sfpt);
     }
+
+    // We only want to use the auto-vectorization check as a trap once per bci. And
+    // PhaseIdealLoop::add_parse_predicate only checks trap limits per method, so
+    // we do a custom check here.
+    if (!C->too_many_traps(cloned_sfpt->jvms()->method(), cloned_sfpt->jvms()->bci(), Deoptimization::Reason_auto_vectorization_check)) {
```

Comment:
Isn't that done by `add_parse_predicate`?

Reply:
@rwestrel I only see `if (!C->too_many_traps(reason)) {` in `PhaseIdealLoop::add_parse_predicate`. And as the comment I put here that only checks the `reason` per `method`, and not per `bci`. Do you see anything else?

Reply:
Seems like it's a bug that `PhaseIdealLoop::add_parse_predicate` doesn't check the `bci` too. Could you fix it?

Reply:
@rwestrel So we would check both, right? But is that what we want for all predicates?

`C->too_many_traps(reason)` checks against `PerMethodTrapLimit`:
```
if (trap_count(reason) >= Deoptimization::per_method_trap_limit(reason)) {
```

But the `bci` check works with `PerBytecodeTrapLimit`, and it actually has a comment like this:
```
if (md->has_trap_at(bci, m, reason) != 0) {
  // Assume PerBytecodeTrapLimit==0, for a more conservative heuristic.
  // Also, if there are multiple reasons, or if there is no per-BCI record,
  // assume the worst.
```
So the `bci` check fails if there has been even a single trapping recorded.

So it seems that such a change would affect the behavior in ways I cannot yet predict.

What do you think?

Reply:
That code is supposed to mirror the `GraphKit::add_parse_predicate()`. It doesn't. Would you like me to fix this separately?

Reply:
> That code

Which code are you referring to?
Ah, probably you are talking about `PhaseIdealLoop::add_parse_predicate`, which is using the method wide check. And `GraphKit::add_parse_predicate` actually queries `GraphKit::too_many_traps`, which knows the current `bci()`, and can query the per-bci count.

> Would you like me to fix this separately?

Yes, please. I definitely don't want to do it in this PR ;)
And I don't have as much experience with traps as you do. We'd have to think a little about what cases this affects, and if performance would go up or down in all those cases.

Reply:
@rwestrel do you consider that a blocking issue for this PR here?

Reply:
No

Reply:
I filed: https://bugs.openjdk.org/browse/JDK-8350330